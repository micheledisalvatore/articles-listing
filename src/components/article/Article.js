import React from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';

import { showArticle } from '../../actions/articles';

import { Content, Img, TextWrapper } from './Article.styled';

export const Article = ({ artwork, content, full, id, title, showArticleAction }) => {
  const handleArticleClick = () => { showArticleAction(id) };

  return (
    <div>
      <Img src={artwork} alt={title} />
      <TextWrapper>
        <h3>{title}</h3>
        <Content dangerouslySetInnerHTML={{__html: content}} onClick={handleArticleClick} full={full} />
      </TextWrapper>
    </div>
  )
}

Article.propTypes = {
  artwork: PropTypes.string.isRequired,
  content: PropTypes.string.isRequired,
  full: PropTypes.bool,
  id: PropTypes.number.isRequired,
  title: PropTypes.string.isRequired,
  showArticleAction: PropTypes.func.isRequired,
}

Article.defaultProps = {
  full: false,
}

export const mapDispatchToProps = (dispatch) => ({
  showArticleAction: bindActionCreators(showArticle, dispatch),
})