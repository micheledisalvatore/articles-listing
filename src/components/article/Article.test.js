/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';

import { Content, Img } from './Article.styled';
import { Article } from './Article';

describe('Given an Article component', () => {
  let component;
  let wrapper;
  const showArticleActionMock = jest.fn();
  const requiredProps = {
    artwork: 'http://foo.bar',
    content: '<p>foo html</p>',
    id: 1,
    title: 'Foo Title',
    showArticleAction: showArticleActionMock,
  };

  describe('when it is rendered', () => {
    beforeEach(() => {
      component = <Article {...requiredProps} />;
      wrapper = shallow(component);
    });

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the Img with the right src attribute', () => {
      expect(wrapper.find(Img)).toHaveProp('src', requiredProps.artwork);
    });

    it('should render the Img with the right alt attribute', () => {
      expect(wrapper.find(Img)).toHaveProp('alt', requiredProps.title);
    });

    it('should render the Content with the full attribute set to false', () => {
      expect(wrapper.find(Content)).toHaveProp('full', false);
    });

    describe('and the user clicks on the Content', () => {
      beforeEach(() => {
        wrapper.find(Content).simulate('click');
      });

      it('should invoke showArticleAction with the right parameter', () => {
        expect(showArticleActionMock).toHaveBeenCalledWith(requiredProps.id);
      })
    })

    describe('and the full property is passed', () => {
      beforeEach(() => {
        wrapper.setProps({
          full: true,
        });
      });

      it('should match the snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should render the Content with the full attribute set to true', () => {
        expect(wrapper.find(Content)).toHaveProp('full', true);
      });
    })
  });
});
