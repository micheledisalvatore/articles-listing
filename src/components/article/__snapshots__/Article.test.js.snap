// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given an Article component when it is rendered and the full property is passed should match the snapshot 1`] = `
<div>
  <styled.img
    alt="Foo Title"
    src="http://foo.bar"
  />
  <styled.div>
    <h3>
      Foo Title
    </h3>
    <styled.button
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>foo html</p>",
        }
      }
      full={true}
      onClick={[Function]}
    />
  </styled.div>
</div>
`;

exports[`Given an Article component when it is rendered should match the snapshot 1`] = `
<div>
  <styled.img
    alt="Foo Title"
    src="http://foo.bar"
  />
  <styled.div>
    <h3>
      Foo Title
    </h3>
    <styled.button
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>foo html</p>",
        }
      }
      full={false}
      onClick={[Function]}
    />
  </styled.div>
</div>
`;
